### ISSUE TITLE
下载完数据，训练结果和文章不一样，效果很差

### ISSUE THREAD

**Issue by zuowanbushiwo** [QingyuLiu0521/ICSD#3]
你好 @QingyuLiu0521 @sssslf 
        我下载了ICSD的数据集， 然后使用   `python train_sed.py --strong_real`  训练，
       训练的结果如下：
```
Epoch 168: 100%|█| 334/334 [00:54<00:00,  6.15it/s, loss=0.348, v_num=0, train/step=46137.0, train/student/tot_self_loss=0.027, traiMetric val/obj_metric improved by 0.000 >= min_delta = 0.0. New best score: 1.899
Epoch 186: 100%|█| 334/334 [00:53<00:00,  6.27it/s, loss=0.207, v_num=0, train/step=51051.0, train/student/tot_self_loss=0.0339, traMetric val/obj_metric improved by 0.000 >= min_delta = 0.0. New best score: 1.900
Epoch 199: 100%|█| 334/334 [00:53<00:00,  6.22it/s, loss=0.132, v_num=0, train/step=54600.0, train/student/tot_self_loss=0.0129, trai
`Trainer.fit` stopped: `max_epochs=200` reached.
Epoch 199: 100%|█| 334/334 [00:53<00:00,  6.20it/s, loss=0.132, v_num=0, train/step=54600.0, train/student/tot_self_loss=0.0129, trai
best model: ./exp/ICSDtask_baseline/version_0/epoch=186-step=51051.ckpt
```

然后使用   测试 ：
```
python train_sed.py --test_from_checkpoint ./exp/ICSDtask_baseline/version_0/epoch=186-step=51051.ckpt
```
结果如下：
```
──────────────────────────────────────────────────────────────────────────────
           Test metric                       DataLoader 0
───────────────────────────────────────────────────────────────────
       /test/tot_energy_kWh              0.001264154794625938
            hp_metric                     0.2532014400110528
   test/student/event_f1_macro           0.43699055910110474
test/student/intersection_f1_macro        0.7744788106630212
     test/student/loss_strong             0.3019263446331024
   test/student/psds1_psds_eval          0.01844333603990256
test/student/psds1_sed_scores_eval       0.03452532851674483
   test/student/psds2_psds_eval           0.2532014400110528
test/student/psds2_sed_scores_eval       0.32577731670686755
   test/teacher/event_f1_macro            0.4312557876110077
test/teacher/intersection_f1_macro        0.794555338952802
     test/teacher/loss_strong            0.29356932640075684
   test/teacher/psds1_psds_eval          0.08378626940858702
test/teacher/psds1_sed_scores_eval       0.09570484034003346
   test/teacher/psds2_psds_eval          0.30224430542847186
test/teacher/psds2_sed_scores_eval        0.3614306664373691
─────────────────────────────────────────────────────────────────────────────────
```

和文章中的结果相差很大，这是为什么？

**Comment by sssslf** [comment]
你好，根据你的情况，判断有可能是sed_scores_eval版本的问题。
我们用的sed_scores_eval版本是0.0.0，在实验的时候没有遇到这个报错，检查了一下我们的lib/python3.8/site-packages/sed_scores_eval/base_modules/ground_truth.py文件，其中没有发现会出现f"No ground truth events for classes {classes_without_ground_truth_events}."这个错误的代码；
我们检查了一下另外版本的sed_scores_eval，发现在0.0.4版本中，/ground_truth.py文件里出现正如你遇到的问题一样，line 70出现f"No ground truth events for classes {classes_without_ground_truth_events}."这个报错代码；
所以目前建议你修改sed_scores_eval版本为0.0.0，再进行实验。

祝一切顺利！

**Comment by zuowanbushiwo** [comment]
你好 @sssslf @QingyuLiu0521 
感谢您的回复。
安装  pip install sed_scores_eval==0.0.0 后确实能运行了

```
       /test/tot_energy_kWh             0.0015615606680512428
            hp_metric                            0.0
   test/student/event_f1_macro           0.23308271169662476
test/student/intersection_f1_macro               0.78
     test/student/loss_strong             0.3106452524662018
   test/student/psds1_psds_eval                  0.0
test/student/psds1_sed_scores_eval               0.0
   test/student/psds2_psds_eval                  0.0
test/student/psds2_sed_scores_eval               0.0
   test/teacher/event_f1_macro            0.2671755850315094
test/teacher/intersection_f1_macro        0.8120805369127517
     test/teacher/loss_strong             0.2998046576976776
   test/teacher/psds1_psds_eval                  0.0
test/teacher/psds1_sed_scores_eval               0.0
   test/teacher/psds2_psds_eval                  0.0
test/teacher/psds2_sed_scores_eval               0.0
```

test/teacher/intersection_f1_macro        0.8120805369127517   就是对应到文章中的   0.8203 吗？

<img width="319" alt="Image" src="https://github.com/user-attachments/assets/a76064f1-54ac-4d77-b5b0-d3fc9d4e40a0" />


我写了一个单独的推理 test程序，处理不同长度的， 从我找的几个数据看效果很差，误报非常的多，不知道我的测试程序对吗？我很仔细的debug 了 程序的推理流程。 

```
import torch
import yaml
import os
import librosa
from torchaudio.transforms import AmplitudeToDB, MelSpectrogram
from desed_task.nnet.CRNN import CRNN
from ICSD_task.local.utils import batched_decode_preds
from desed_task.utils.encoder import ManyHotEncoder
from ICSD_task.local.classes_dict import classes_labels

def take_log(mels):
    amp_to_db = AmplitudeToDB(stype="amplitude")
    amp_to_db.amin = 1e-5  
    return amp_to_db(mels).clamp(min=-50, max=80)  # clamp to reproduce old code

def find_wav_files(directory):
    wav_files = []
    for root, dirs, files in os.walk(directory):
        for file in files:
            if file.lower().endswith('.wav'):
                full_path = os.path.join(root, file)
                wav_files.append(full_path)
    return wav_files

def filter_state_dict(raw_dict, prefix):
    return {
        key.replace(prefix, ""): value
        for key, value in raw_dict.items()
        if key.startswith(prefix) 
        and not ("total_ops" in key or "total_params" in key)  
    }



test_state_dict = torch.load('/home/yang/ICSD/ICSD_task/exp/ICSDtask_baseline/version_0/epoch=186-step=51051.ckpt')["state_dict"]

student_state_dict = filter_state_dict(test_state_dict, "sed_student.")

teacher_state_dict = filter_state_dict(test_state_dict, "sed_teacher.")


with open('/home/yang/ICSD/ICSD_task/exp/ICSDtask_baseline/version_0/hparams.yaml', "r") as f:
    configs = yaml.safe_load(f)

sed_student = CRNN(**configs["net"])
sed_teacher = CRNN(**configs["net"])

sed_student.load_state_dict(student_state_dict)
sed_teacher.load_state_dict(teacher_state_dict)

feat_params = configs["feats"]
mel_spec = MelSpectrogram(
    sample_rate=feat_params["sample_rate"],
    n_fft=feat_params["n_window"],
    win_length=feat_params["n_window"],
    hop_length=feat_params["hop_length"],
    f_min=feat_params["f_min"],
    f_max=feat_params["f_max"],
    n_mels=feat_params["n_mels"],
    window_fn=torch.hamming_window,
    wkwargs={"periodic": False},
    power=1,
)


sed_student.eval()
sed_teacher.eval()

wav_files_list = find_wav_files('/home/yang/baby_cry')
for wav_file in wav_files_list:
    (waveform, _) = librosa.core.load(wav_file, sr=feat_params["sample_rate"], mono=True)
    waveform = torch.from_numpy(waveform[None, :])
    waveform_len = waveform.shape[1]

    audio_len = waveform_len / feat_params["sample_rate"]
    encoder = ManyHotEncoder(
            list(classes_labels.keys()),
            audio_len= audio_len,
            frame_len=configs["feats"]["n_filters"],
            frame_hop=configs["feats"]["hop_length"],
            net_pooling=configs["data"]["net_subsample"],
            fs=configs["data"]["fs"],
        )

    mels = mel_spec(waveform)
    log_mels = take_log(mels)
    scaler_out = (log_mels - torch.amin(log_mels, dim=[1,2], keepdim=True)) / (
        torch.amax(log_mels, dim=[1,2], keepdim=True)
        - torch.amin(log_mels, dim=[1,2], keepdim=True)
        + 1e-8
    )


    with torch.no_grad():
        strong_preds_student, week_preds_student = sed_student(scaler_out)

    with torch.no_grad():
        strong_preds_teacher, week_preds_teacher = sed_teacher(scaler_out)


    # (
    #     scores_raw_student_strong, scores_postprocessed_student_strong,
    #     decoded_student_strong,
    # ) = batched_decode_preds(
    #     strong_preds_student,
    #     wav_file,
    #     encoder,
    #     median_filter=configs["training"]["median_window"],
    #     thresholds=(0.2, 0.3, 0.5),
    # )
    # for th in decoded_student_strong:
    #     df = decoded_student_strong[th] 
    #     df = df.drop(columns=['filename'])
    #     filtered_df = df[df['event_label'] == 'Infantcry']
    #     baby_count = len(filtered_df)
    #     if(baby_count != 0 and "negative" in wav_file) :
    #         print(f"False Positive detected {wav_file} events using threshold {th}:")
    #     elif (baby_count == 0 and "positive" in wav_file): 
    #         print(f"True Negative  detected {wav_file} events using threshold {th}:")   

    (
        scores_raw_teacher_strong, scores_postprocessed_teacher_strong,
        decoded_teacher_strong,
    ) = batched_decode_preds(
        strong_preds_teacher,
        wav_file,
        encoder,
        median_filter=configs["training"]["median_window"],
        thresholds=(0.7,0.8,0.9),
    )

    for th in decoded_teacher_strong:
        df = decoded_teacher_strong[th] 
        df = df.drop(columns=['filename'])
        filtered_df = df[df['event_label'] == 'Infantcry']
        baby_count = len(filtered_df)
        if(baby_count != 0 and "negative" in wav_file) :
            print(f"False Positive detected {wav_file} events using threshold {th}:")
        elif (baby_count == 0 and "positive" in wav_file): 
            print(f"True Negative  detected {wav_file} events using threshold {th}:") 

print("Hello World")
```
期待你们的指导，谢谢！

**Comment by QingyuLiu0521** [comment]
代码修改了，可以重新git clone一下，现在只显示segment_f1和intersection_f1
```
───────────────────────────────────────────────────────────────────
           Test metric                       DataLoader 0
───────────────────────────────────────────────────────────────────
       /test/tot_energy_kWh              0.004576365929096937
            hp_metric                     0.875787854213072
test/student/intersection_f1_macro        0.9224777775310449
     test/student/loss_strong            0.11429200321435928
    test/student/seg_f1_macro             0.9199017286300659
test/teacher/intersection_f1_macro        0.9284161962363571
     test/teacher/loss_strong             0.1182742491364479
    test/teacher/seg_f1_macro             0.9237219095230103
───────────────────────────────────────────────────────────────────
```
关于使用huggingface上的ckpt测试得到的结果和论文不一致的问题，可以重新用huggingface上新上传的https://huggingface.co/datasets/QingyuLiu1/ICSD/blob/main/hybrid_baseline.ckpt测一下，应该能得到和论文里一致的结果

**Comment by zuowanbushiwo** [comment]
感谢回复，使用最新的代码 结合 hybrid_baseline.ckpt 文件 能得到 论文的数据了。

```
       /test/tot_energy_kWh              0.002870689146220684
            hp_metric                            0.0
test/student/intersection_f1_macro        0.8203389644622803
     test/student/loss_strong            0.32783907651901245
    test/student/seg_f1_macro             0.8496732115745544
test/teacher/intersection_f1_macro        0.8243243098258972
     test/teacher/loss_strong             0.3160060942173004
    test/teacher/seg_f1_macro             0.8605042099952698

```

我使用 hybrid_baseline.ckpt 测试，召回率比较高，精确率很低。只要有高音调的基本都会误检，像唱歌，欢呼声，更不用说男人, 女人的哭声，婴儿的其它声音 基本也都会误检，

**Comment by zuowanbushiwo** [comment]
@QingyuLiu0521  @sssslf 
后来仔细看你们的文章，有一些疑问：
对于  Competitive CRNN+BEATs 的 数据准备 过程：
  (1)    从 Figure 5 看，week label 的数据 保持不变，还是只有 2类，只有 Infantcry 和  Snoring 两类？
  (2)    对于strong label 数据 新增了 3类，但是对于 data/metadata/train/real_strong_train.tsv  中标签不做改变， 比如       
```
      strong_Infantcry_1.wav	3.807	6.338	Infantcry
      strong_Infantcry_2.wav	7.91	9.65	Infantcry 
```
对于 strong_Infantcry_1.wav 文件，即使 在其它时间 (0,  3.8) 有speech 或者 music  也不修改？ 同时 对于  (3.807, 6.338) 即使 有speech 或者 music  也不修改？因为audioset 中的标签是可以同时发生的，如果上面的  strong_Infantcry_1.wav 来自 audioset ，那就可能同时发生。

  (3) 对于  MT-CRNN 有没有做 Competitive CRNN+BEATs 和 CRNN+BEATs  的这种实验，我看文章没有提到？结果会像  CRNN+BEATs 一样有提升吗？

谢谢！
